### PR TITLE
Add field selection rendering of allowedProjects for BitbucketRepoPicker

### PR DESCRIPTION
**Pull request by akundrock**
Signed-off-by: Andrew Kundrock <akundrock@somalogic.com>

## Hey, I just made a Pull Request!

Add allowedProjects to the repoUrlPicker/BitbucketRepoPicker schema. This allows a developer to specify which projects are allowed in Bitbucket Cloud. I find this to be helpful to control where a repo can be selected via dropdown list.

Projects only seem to be used by Bitbucket Cloud and needed to be added to the ZOD schema

image

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [X] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by akundrock** [comment]
> ## Changed Packages
> Package Name	Package Path	Changeset Bump	Current Version
> @backstage/plugin-scaffolder-backend	plugins/scaffolder-backend	**minor**	`v1.8.1-next.1`
> @backstage/plugin-scaffolder	plugins/scaffolder	**minor**	`v1.9.0-next.1`

The scaffolder backend has no code changes but I updated the example Bitbucket scaffolder action to include the new ui:options allowedProjects.
Not sure if I should have tagged that package for a minor update

File: plugins/scaffolder-backend/sample-templates/bitbucket-demo/template.yaml [hunk 1]
```diff
@@ -20,7 +20,15 @@ spec:
           ui:options:
             allowedHosts:
               - bitbucket.org
-              - server.bitbucket.com
+            allowedOwners:
```
Comment (agentbellnorm) [inline plugins/scaffolder-backend/sample-templates/bitbucket-demo/template.yaml:23]: I'm not sure of the best place, probably a comment or some documentation file, but I feel like we should mention somewhere that these lists are optional, and what happens when you don't provide them. What do you think?

**Comment by akundrock** [comment]
Added test coverage and inline descriptors around the Bitbucket-specific exports for the repoUrlPicker.

Pulled master to make sure everything was up to date and ended up with ~420 commits, a bunch of Github tags, and extra reviewers.

Sorry, everyone!
If this was a mistake I can cherry-pick my work against backstage's current master

**Comment by agentbellnorm** [comment]
@akundrock Hehe, oops! Not sure what happened, but yes maybe it's easiest to cherry pick your commits, create a new clean PR and close this one?

File: .changeset/wet-tips-yawn.md [hunk 2]
```diff
@@ -3,4 +3,8 @@
 '@backstage/plugin-scaffolder-backend': minor
 ---
 
-Add field selection rendering of allowed Projects for BitbucketRepoPicker. Allows a user to define allowed projects via the ui:field RepoUrlPicker ui:options allowed Projects
+Add field selection rendering of allowed Projects for BitbucketRepoPicker. Allows a user to define allowed projects via the ui:field RepoUrlPicker ui:options allowed Projects.
+
+Add comments about expected values for BitbucketRepoPicker: Projects and workspaces
+
+Update tests to cover null, empty list, and list values for allowed Bitbucket projects
```
Comment (agentbellnorm) [inline .changeset/wet-tips-yawn.md:10]: This text is what will go into the changelog, so we don't need to specify details of each commit here. We only need to say what the change is doing. I think the first line you have is enough:
`Add field selection rendering of allowed Projects for BitbucketRepoPicker. Allows a user to define allowed projects via the ui:field RepoUrlPicker ui:options allowed Projects.`

Actually I would probably word it a bit differently:
`Enable stricter selection of allowed Projects for BitbucketRepoPicker. If enabled, a user can only choose from a list of allowed projects from the ui:field RepoUrlPicker ui:options.allowedProjects`

**Comment by camilaibs** [comment]
> @akundrock Hehe, oops! Not sure what happened, but yes maybe it's easiest to cherry pick your commits, create a new clean PR and close this one?

I rebased this branch with master and that fixed it, could you check if everything is correct after rebasing @akundrock ?

File: plugins/scaffolder/src/components/fields/RepoUrlPicker/BitbucketRepoPicker.tsx [hunk 3]
```diff
@@ -21,17 +21,36 @@ import InputLabel from '@material-ui/core/InputLabel';
 import { Select, SelectItem } from '@backstage/core-components';
 import { RepoUrlPickerState } from './types';
 
+/**
+ * The underlying component that is rendered in the form for the `BitbucketRepoPicker`
+ * field extension.
+ *
+ * @public
+ * @param allowedOwners - Allowed workspaces for the Bitbucket cloud repository
```
Comment (adamdmharvey) [inline plugins/scaffolder/src/components/fields/RepoUrlPicker/BitbucketRepoPicker.tsx:29]: allowed WORKSPACES, or *owners* of workspaces?  wonder if that should be clarified? (though to be fair, I'm not an active user of BitBucket, so defer to your expertise :) )

File: .changeset/wet-tips-yawn.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+---
+'@backstage/plugin-scaffolder': minor
+'@backstage/plugin-scaffolder-backend': minor
+---
+
+Add field selection rendering of allowed Projects for BitbucketRepoPicker. Allows a user to define allowed projects via the ui:field RepoUrlPicker ui:options allowed Projects.
+
+Add comments about expected values for BitbucketRepoPicker: Projects and workspaces
+
+Update tests to cover null, empty list, and list values for allowed Bitbucket projects
```
Comment (Rugvip) [inline .changeset/wet-tips-yawn.md:10]:
```suggestion
The `RepoUrlPicker` field extension now has an `allowedProjects` option for narrowing the selection of Bitbucket URLs.
```

`BitbucketRepoPicker` is actually an internal concern too, I shortened it up a bit more x)

File: .changeset/wet-tips-yawn.md [hunk 5]
```diff
@@ -0,0 +1,10 @@
+---
+'@backstage/plugin-scaffolder': minor
+'@backstage/plugin-scaffolder-backend': minor
```
Comment (Rugvip) [inline .changeset/wet-tips-yawn.md:3]:
```suggestion
'@backstage/plugin-scaffolder': minor
```

The sample template isn't part of the published package, so no need for a changeset

**Review by Rugvip — APPROVED** [review]
Nice, thank you! 👍 

Gonna apply some changeset updates myself so that we can ship it in todays release :>

**Comment by akundrock** [comment]
Thanks, everyone for cleaning up my comments and the strange pull from master! Glad this change could make it into the release and I have plenty of ideas for future commits!

**Comment by ronabet** [comment]
Hi, is this feature active? 
Can I install the last versions of @backstage/plugin-scaffolder-backend and @backstage/plugin-scaffolder?

**Comment by ronabet** [comment]
> 

Hi, thanks for this feature.
When can I use it? and what should I do to have it?

**Pull request merged**